evennia/contrib/aisystem/README.md

```diff
@@ -0,0 +1,661 @@
+method's own arguments list.
+
+`open(self, bb)` - this method sets up a node instance that did not return
+RUNNING during the previous iteration of the AI. It is called immediately before
+`update(bb)`. You can use it to, for instance, reset the node instance's
+node-specific data. 
+
+`close(self, bb)` - this method cleans up a node that did not return RUNNING
+during the previous iteration of the AI. It is called after the AI has finished
+iterating through the tree.
+
+`on_add_child(self, node)` - called after a node (specified in the arguments
+list) is set as this node's child or added to this node's list of children.
+
+`on_remove_child(self, node)` - called after a node (specified in the arguments
+list) is no longer this node's child or is removed from this node's list of
+children.
+
+`on_blackboard_setup(self, bb, override=False)` - called whenever the blackboard
+associated with an instance of this node is set up. The override argument
+supplied is the same one provided by the @aisetup command (if no override
+argument is provided to that command, the override argument for
+on\_blackboard\_setup is set to False, else it is True).
+
+You should ideally set blackboard data within the on\_blackboard\_setup method
+using the following pattern:
+```
+    if not bb['nodes'][self.hash].has_key(key_name) or override:
+        bb['nodes'][self.hash][key_name] = value
+
+    if not bb['globals'].has_key(globals_key_name) or override:
+        bb['globals'][globals_key_name] = globals_value
+```
+
+The following blackboard data is available:
+
+`bb['agent']` - the agent that holds the blackboard
+
+`bb['tree']` - the agent's origin tree
+
+`bb['nodes'][some_node.hash]` - the dictionary of node-specific data for
+                                `some_node`
+
+`bb['globals']` - the dictionary of the agent's global data
+
+For more information on the entries of the latter two dictionaries, see the
+"Blackboards" section.
+
+Nodes automatically handle exceptions encountered within the `update` method by
+storing the error message and returning the ERROR status. Ultimately, if the
+nodes you have coded are set up properly (i.e. they return ERROR whenever one of
+their children returns ERROR) the entire tree will return the ERROR status, so
+you can check for that status within your scripts or tickerhandlers and display
+that error message accordingly. Errors are stored in the dictionary
+`bb['globals']['errors']`, whose keys are the full hash values of the nodes
+that have encountered errors. To add an error string to the dictionary, do this:
+
+```
+bb['globals']['errors'][node.hash] = error_string
+```
+
+The tree does not clean up its error strings automatically. To do so, you should
+run the agent's `ai.clean()` method:
+
+```
+agent.ai.clean()
+```
+
+Finally, you should never need to overload the node's \_\_init\_\_ method.
+
+### Limitations
+
+The theoretical upper limit for the number of nodes in a tree is 238,328.
```

Comment:
I assume this is due to hash clashes, but it is best to explain this explicitly.